### PR TITLE
improve command line argument validation

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- Validate that the number of values passed via `--gridDist, --grid, ...` is not more than allowed.
- Validate the correctness of particle boundary settings only once.
- Move the help for particle boundaries, pml, ... behind the basic parameters e.g. `--devices`, ...

This PR helps to avoid issues like https://github.com/ComputationalRadiationPhysics/picongpu/issues/4446

Error message example:
```
Unhandled exception of type 'St13runtime_error' with message 'expression (offsets().size() <= 3) 
failed in file (picongpu/include/picongpu/../picongpu/simulation/stage/ParticleBoundaries.hpp:108) : 
Invalid number of particle boundary offsets for species 'ec'.', terminating
```

### CODE REVIEW

File: include/picongpu/simulation/control/Simulation.hpp [hunk 1]
```diff
@@ -166,26 +168,28 @@ namespace picongpu
...
+
             // check on correct number of devices. fill with default value 1 for missing dimensions
-            if(devices.size() > 3)
-            {
-                std::cerr << "Invalid number of devices.\nuse [-d dx=1 dy=1 dz=1]" << std::endl;
-            }
-            else
-                while(devices.size() < 3)
-                    devices.push_back(1);
+            while(devices.size() < 3)
```
Comment (steindev) [inline include/picongpu/simulation/control/Simulation.hpp:177]: `while()` here but `if(` in line 192. Homogenize?

File: include/picongpu/simulation/control/Simulation.hpp [hunk 2]
```diff
@@ -218,6 +222,8 @@ namespace picongpu
                 }
             }
 
+            PMACC_VERIFY_MSG(gridDistribution.size() <= 3, "Invalid number of grid distributions.");
```
Comment (steindev) [inline include/picongpu/simulation/control/Simulation.hpp]: The other asserts have a comment how to use the program option correctly, see e.g. line 187. You think it is not necessary here?

File: include/picongpu/simulation/stage/ParticleBoundaries.hpp [hunk 3]
```diff
@@ -98,54 +98,77 @@ namespace picongpu
                      */
                     void operator()()
                     {
-                        for(uint32_t d = 0; d < simDim; d++)
+                        static bool validateUserInputs = true;
+                        if(validateUserInputs)
```
Comment (steindev) [inline include/picongpu/simulation/stage/ParticleBoundaries.hpp:102]: Why encapsulate this in an `if()`? The condition is always true anyways.

File: include/picongpu/simulation/stage/ParticleBoundaries.hpp [hunk 4]
```diff
@@ -98,54 +98,77 @@ namespace picongpu
...
+                                {
+                                    if(T_Species::boundaryDescription()[d].kind == particles::boundary::Kind::Periodic)
+                                        throw std::runtime_error(
+                                            "Boundary kind " + errorString
+                                            + " is not compatible with --periodic value");
+                                    T_Species::boundaryDescription()[d].kind = particles::boundary::Kind::Reflecting;
+                                }
+                                if(kindName == "thermal")
+                                {
+                                    if(T_Species::boundaryDescription()[d].kind == particles::boundary::Kind::Periodic)
```
Comment (steindev) [inline include/picongpu/simulation/stage/ParticleBoundaries.hpp:144]: There is something wrong, either here in the condition, her in the error message, or in the check above.

File: include/picongpu/simulation/stage/ParticleBoundaries.hpp [hunk 5]
```diff
@@ -98,54 +98,77 @@ namespace picongpu
...
+                                            + " is not compatible with --periodic value");
+                                }
+                                if(kindName == "absorbing")
+                                {
+                                    // For now it must match the default-set boundary kind
+                                    if(T_Species::boundaryDescription()[d].kind
+                                       != particles::boundary::Kind::Absorbing)
+                                        throw std::runtime_error(
+                                            "Boundary kind " + errorString
+                                            + " is not compatible with --periodic value");
```
Comment (steindev) [inline include/picongpu/simulation/stage/ParticleBoundaries.hpp:165]: Again, I believe either here or in the check above something is wrong. Probably the error message.
Isn't it necessary to set the `T_Species::boundaryDescription()[d].kind` here or in the check above as well?

**Review by steindev — COMMENTED** [review]
I have some questions. It is hard to follow the logic in the checks performed in `ParticleBoundaries.hpp`

File: include/picongpu/simulation/stage/ParticleBoundaries.hpp (hunk 3 above)
Comment (psychocoderHPC) [reply include/picongpu/simulation/stage/ParticleBoundaries.hpp:102]: there is no need to validate it more than once

File: include/picongpu/simulation/stage/ParticleBoundaries.hpp (hunk 5 above)
Comment (psychocoderHPC) [reply include/picongpu/simulation/stage/ParticleBoundaries.hpp:165]: What should be wrong here? I have not changed anything in this PR.

File: include/picongpu/simulation/control/Simulation.hpp (hunk 1 above)
Comment (psychocoderHPC) [reply include/picongpu/simulation/control/Simulation.hpp:177]: Could you explain what you see here as a problem? This line is only filling up the vector `devices` with `1`
Line 192 is extending the `gridSize` by `1` in case we run a 2D simulation.

File: include/picongpu/simulation/control/Simulation.hpp (hunk 2 above)
Comment (psychocoderHPC) [reply include/picongpu/simulation/control/Simulation.hpp]: In that case, I would not try to guide the user because there are many ways to user grid distributions.

File: include/picongpu/simulation/stage/ParticleBoundaries.hpp (hunk 3 above)
Comment (psychocoderHPC) [reply include/picongpu/simulation/stage/ParticleBoundaries.hpp:102]: The condition is true for the first evaluation and then set to false and is not evaluated a second time.

File: include/picongpu/simulation/stage/ParticleBoundaries.hpp (hunk 4 above)
Comment (psychocoderHPC) [reply include/picongpu/simulation/stage/ParticleBoundaries.hpp:144]: What is wrong, if so we need to handle it in a separate PR, there are no changes in these lines only the indention is changed because of the additional if condition.

File: include/picongpu/simulation/control/Simulation.hpp (hunk 2 above)
Comment (psychocoderHPC) [reply include/picongpu/simulation/control/Simulation.hpp]: changed to `"Too many grid distribution directions given.  A maximum of three directions are supported."`

File: include/picongpu/fields/incidentField/profiles/Free.def [hunk 6]
```diff
@@ -30,7 +30,7 @@ namespace picongpu
         {
             namespace profiles
             {
-                /** Free profile tag using given functors
+                /** Free profile tag usingInvalid number of grid distributions. given functors
```
Comment (PrometheusPi) [inline include/picongpu/fields/incidentField/profiles/Free.def]: I do not understand that comment/description - but this is fine

**Review by PrometheusPi — DISMISSED** [review]
@steindev you had a number of comments. What's your current status on this pull request?

File: include/picongpu/fields/incidentField/profiles/Free.def (hunk 6 above)
Comment (psychocoderHPC) [reply include/picongpu/fields/incidentField/profiles/Free.def]: This should not be part of the PR, I think I was in the wrong window of my IDE and past it in.

File: include/picongpu/simulation/stage/ParticleBoundaries.hpp (hunk 5 above)
Comment (PrometheusPi) [reply include/picongpu/simulation/stage/ParticleBoundaries.hpp:165]: If there is an error, we need a separate PR as suggested by @psychocoderHPC

**Review by PrometheusPi — APPROVED** [review]
